File: packages/sdk-python/korveo/queue.py
import asyncio
from typing import List, Optional

from .span import Span


class BoundedQueue:
    """Asyncio-backed bounded FIFO queue. Drops new items on overflow."""

    def __init__(self, max_size: int = 10_000):
        self._queue: asyncio.Queue = asyncio.Queue(maxsize=max_size)
        self._dropped = 0

    def put_nowait(self, span: Span) -> bool:
        try:
            self._queue.put_nowait(span)
            return True
        except asyncio.QueueFull:
            self._dropped += 1
            return False

    async def drain(self, max_items: Optional[int] = None) -> List[Span]:
        items: List[Span] = []
        limit = max_items if max_items is not None else self._queue.qsize()
        for _ in range(limit):
            try:
                items.append(self._queue.get_nowait())
            except asyncio.QueueEmpty:
                break
        return items

    def __len__(self) -> int:
        return self._queue.qsize()

    @property
    def dropped(self) -> int:
        return self._dropped

File: packages/sdk-typescript/src/queue.ts
import { Span } from './span.js';

/** Bounded FIFO queue. Drops new items on overflow — never throws. */
export class BoundedQueue {
  private readonly items: Span[] = [];
  private droppedCount = 0;

  constructor(private readonly maxSize: number) {}

  put(span: Span): boolean {
    if (this.items.length >= this.maxSize) {
      this.droppedCount++;
      return false;
    }
    this.items.push(span);
    return true;
  }

  drain(maxItems?: number): Span[] {
    const limit = maxItems === undefined ? this.items.length : maxItems;
    return this.items.splice(0, limit);
  }

  get size(): number {
    return this.items.length;
  }

  get dropped(): number {
    return this.droppedCount;
  }
}
